common/hashtable_key_context.h

```diff
@@ -69,16 +97,138 @@ namespace Carbon {
+  // overload is provided at all.
+  template <typename TranslateKeyT>
+  auto TranslateKey(const TranslateKeyT& /*key*/) const -> int {
+    // A static_assert that will fail on any actual instantiation (it can't be
+    // instantiated with a void type). We have to make this dependent as
+    // Clang-16 will fail to compile even when the definition is never
+    // instantiated otherwise.
+    static_assert(
+        std::same_as<TranslateKeyT, void>,
+        "No `TranslateKey` overload was provided by the derived type!");
+  }
+
+  template <typename AnyKeyT>
+  auto HashKey(const AnyKeyT& key, uint64_t seed) const -> HashCode;
+
+  template <typename AnyKeyT, typename TableKeyT>
+  auto KeyEq(const AnyKeyT& lhs_key, const TableKeyT& rhs_key) const -> bool;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// Only implementation details below this point.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+namespace InternalHashtableEqDispatch {
+
+inline auto CarbonHashtableEq(const llvm::APInt& lhs, const llvm::APInt& rhs)
+    -> bool {
+  return lhs.getBitWidth() == rhs.getBitWidth() && lhs == rhs;
+}
+
+inline auto CarbonHashtableEq(const llvm::APFloat& lhs,
+                              const llvm::APFloat& rhs) -> bool {
+  return lhs.bitwiseIsEqual(rhs);
+}
+
+template <typename LeftT, typename RightT>
+inline auto CarbonHashtableEq(const LeftT& lhs, const RightT& rhs) -> bool
+  requires(requires {
+    { lhs == rhs } -> std::convertible_to<bool>;
+  })
+{
+  return lhs == rhs;
+}
+
+template <typename LeftT, typename RightT>
+inline auto DispatchImpl(const LeftT& lhs, const RightT& rhs) -> bool {
+  // This unqualified call will find both the overloads in our internal
+  // namespace above and ADL-found functions within an associated namespace for
+  // either `LeftT` or `RightT`.
+  return CarbonHashtableEq(lhs, rhs);
+}
+
+}  // namespace InternalHashtableEqDispatch
+
+template <typename LeftT, typename RightT>
+inline auto HashtableEq(const LeftT& lhs, const RightT& rhs) -> bool {
+  return InternalHashtableEqDispatch::DispatchImpl(lhs, rhs);
+}
+
+template <typename AnyKeyT>
+auto DefaultKeyContext::HashKey(const AnyKeyT& key, uint64_t seed) const
+    -> HashCode {
+  return HashValue(key, seed);
+}
+
+template <typename AnyKeyT, typename TableKeyT>
+auto DefaultKeyContext::KeyEq(const AnyKeyT& lhs_key,
+                              const TableKeyT& rhs_key) const -> bool {
+  return HashtableEq(lhs_key, rhs_key);
+}
+
+template <typename DerivedT>
+template <typename AnyKeyT>
+auto TranslatingKeyContext<DerivedT>::HashKey(const AnyKeyT& key,
+                                              uint64_t seed) const -> HashCode {
+  const DerivedT& self = *static_cast<const DerivedT*>(this);
+  if constexpr (requires { self.TranslateKey(key); }) {
+    return HashValue(self.TranslateKey(key), seed);
+  } else {
     return HashValue(key, seed);
   }
```

Comment:
I'm a bit worried that accidentally passing in an unexpected / bogus (but hashable) type as the key type will silently "work" here. I suppose it's much less likely for the `KeyEq` code below to compile, though, so perhaps I shouldn't worry so much.

I'm wondering if we could do something to remove the risk: for example, perhaps `TranslatingKeyContext` could be given a list of key types that don't need translating, and could reject keys that aren't in the list but can't be translated. (It looks like we don't currently have an example with more than one untranslated key type, so maybe we could use `DerivedT::Key` as the untranslated type?)

Reply:
The `DefaultKeyContext` lets you just hash any type, and so it seemed strange to restrict things in this helper.

I'm also worried about listing the types being onerous because we won't get implicit conversions or other things here. If you *don't* want to translate a `ssize_t` key type, it'll be awkward to call things with integer literals because they won't promote. Even things like `nullptr` make it tricky to list all expected types.

That's part of why I used the "if the call is valid, do the call" model above.

Reply:
OK. I'm OK with moving ahead with this. Let's just keep an eye out -- if we see this kind of wrong-key-type problem coming up and being surprising, we can reconsider adding some kind of guard rail. I *think* we'll pretty consistently get a compile-time error in the key comparison rather than something mysterious where lookups fail because we use an inconsistent hash value, though.